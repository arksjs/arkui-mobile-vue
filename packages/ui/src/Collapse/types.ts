import type { IconData } from '../Icon/types'

export type ActiveName = string | number

export type OnChange = (activeNames: ActiveName[]) => void

export type ItemOnToggle = (payload: {
  name: ActiveName
  spread: boolean
}) => void

export interface CollapseProps {
  activeNames?: ActiveName | ActiveName[]
  accordion?: boolean
}

export interface CollapseEmits {
  onUpdateActiveNames?: OnChange
  onChange?: OnChange
}

export interface CollapseItemProps {
  icon?: IconData
  title?: string
  name: string | number
  disabled?: boolean
}

export interface CollapseItemEmits {
  onToggle?: ItemOnToggle
}
