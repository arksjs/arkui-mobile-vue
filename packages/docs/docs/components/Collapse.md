# Collapse/CollapseItem 折叠面板

<CodeDemo name="Collapse">

<<< @/../../demo/src/components/Show/Collapse/index.vue

</CodeDemo>

## Import

```js
import { AkCollapse, AkCollapseItem } from 'arkui-mobile-vue'
```

具体的引入方式可以参考[引入组件](../guide/import.md)。

### Import Type

组件导出的类型定义：

```ts
import type {
  CollapseActiveName,
  CollapseOnChange,
  CollapseItemOnToggle
} from 'arkui-mobile-vue'
```

## Collapse Props

| 属性                 | 类型                 | 默认值 | 必填 | 说明                                                                    |
| -------------------- | -------------------- | ------ | ---- | ----------------------------------------------------------------------- |
| v-model:active-names | (string \| number)[] | []     | 否   | 当前展开面板的 `name` 列表，需要每个 `CollapseItem` 组件都设置上 `name` |
| accordion            | boolean              | false  | 否   | 是否开启手风琴模式                                                      |

## Collapse Events

| 事件   | 描述           | 回调函数参数                      | TypeScript 函数  |
| ------ | -------------- | --------------------------------- | ---------------- |
| change | 切换面板时触发 | activeNames: (string \| number)[] | CollapseOnChange |

## Collapse Slots

### #default

注：其中只可放置 [CollapseItem](./Collapse.md#collapseitem-折叠面板子项) 组件，否则会导致未定义的行为。

```vue
<ak-collapse>
  <ak-collapse-item title="标题1" name="row1">
    <div class="pad">
    代码是写出来给人看的，附带能在机器上运行
    </div>
  </ak-collapse-item>
  <ak-collapse-item title="标题2" name="row2">
    <div class="pad">
    代码是写出来给人看的，附带能在机器上运行
    </div>
  </ak-collapse-item>
  <ak-collapse-item title="标题3" name="row3">
    <div class="pad">
    代码是写出来给人看的，附带能在机器上运行
    </div>
  </ak-collapse-item>
</ak-collapse>
```

## CollapseItem Props

| 属性     | 类型    | 默认值 | 必填 | 说明                                                 |
| -------- | ------- | ------ | ---- | ---------------------------------------------------- |
| title    | string  | ''     | 否   | 面板子项标题                                         |
| icon     | string  |        | 否   | 面板子项标题前图标，使用 [Icon](./Icon.md) 组件      |
| name     | string  |        | 否   | 唯一标识，设置后可以配合 `v-model:active-names` 使用 |
| disabled | boolean | false  | 否   | 是否禁用                                             |

## CollapseItem Events

| 事件   | 描述                    | 回调函数参数                                                                                 | TypeScript 函数      |
| ------ | ----------------------- | -------------------------------------------------------------------------------------------- | -------------------- |
| toggle | 面板子项展开/收起时触发 | payload: { name: string \| number, spread: boolean } 其中 spread=true 展开/spread=false 收起 | CollapseItemOnToggle |
