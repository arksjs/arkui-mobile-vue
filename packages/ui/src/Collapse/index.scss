@import '../style/var.scss';

.ta-collapse-item {
  &_body {
    overflow: hidden;
    transition: height 0.2s;
    will-change: height;

    &::before {
      content: '';
    }
  }

  &_content {
    font-size: 14px;
    color: $font-color;
    line-height: 20px;
  }

  + .ta-collapse-item::before {
    content: '';
  }

  &_header {
    &:not(:last-child)::after {
      content: none;
    }

    .ta-cell_link-icon {
      transition: transform 0.2s;
    }
  }
}
