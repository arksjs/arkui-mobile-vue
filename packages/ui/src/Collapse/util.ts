export const getItemClasses = (active: boolean) => [
  'ak-collapse-item',
  'ak-horizontal-hairline',
  { active }
]
