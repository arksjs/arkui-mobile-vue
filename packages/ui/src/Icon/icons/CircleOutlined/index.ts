export { default as default } from './CircleOutlined.vue'
