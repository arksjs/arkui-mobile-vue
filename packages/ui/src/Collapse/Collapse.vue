<template>
  <div class="ak-collapse">
    <slot></slot>
  </div>
</template>

<script lang="ts">
import { defineComponent, onMounted, watch, provide } from 'vue'
import type { PropType } from 'vue'
import { cloneData, isSameArray, isStringNumberMixArray } from '../helpers/util'
import { stringNumberArrayMixValidator } from '../helpers/validator'
import { useGroup } from '../hooks/use-group'
import type { ActiveName, CollapseEmits } from './types'
import type { PropsToEmits } from '../helpers/types'

export default defineComponent({
  name: 'ak-collapse',
  props: {
    activeNames: {
      type: [Number, String, Array] as PropType<ActiveName | ActiveName[]>,
      validator: stringNumberArrayMixValidator,
      default: () => [] as ActiveName[]
    },
    accordion: {
      type: Boolean,
      default: false
    }
  },
  emits: {
    'update:activeNames': payload => isStringNumberMixArray(payload),
    change: payload => isStringNumberMixArray(payload)
  } as PropsToEmits<CollapseEmits>,
  setup(props, { emit }) {
    let activeNames2: ActiveName[] = []

    const { children } = useGroup('collapse')

    function updateValue(val: ActiveName | ActiveName[]) {
      let values = cloneData(
        stringNumberArrayMixValidator(val)
          ? Array.isArray(val)
            ? val
            : [val]
          : []
      )

      if (props.accordion) {
        // 手风琴模式只保留一个值
        values = values.slice(0, 1)
      }

      if (Array.isArray(values) && isSameArray(values, activeNames2)) {
        return
      }

      activeNames2 = []

      children.forEach(child => {
        const childName = child.getName() as ActiveName

        if (childName && values.includes(childName)) {
          activeNames2.push(childName)
          child.show()
        } else {
          child.hide()
        }
      })
    }

    function onChange(uid: number) {
      activeNames2 = []

      if (props.accordion) {
        children.forEach(child => {
          if (child.uid === uid) {
            child.getActive() &&
              child.getName() &&
              activeNames2.push(child.getName())
          } else {
            child.hide()
          }
        })
      } else {
        children.forEach(child => {
          child.getActive() &&
            child.getName() &&
            activeNames2.push(child.getName())
        })
      }

      emit('update:activeNames', cloneData(activeNames2))
      emit('change', cloneData(activeNames2))
    }

    onMounted(() => updateValue(props.activeNames))

    watch(
      () => props.activeNames,
      val => updateValue(val)
    )

    provide('akCollapseChange', onChange)

    return {}
  }
})
</script>
