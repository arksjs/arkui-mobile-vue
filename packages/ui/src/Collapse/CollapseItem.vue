<template>
  <div :class="classes">
    <Cell
      class="ak-collapse-item_header"
      :label="title"
      :icon="icon"
      :disabled="disabled"
      isLink
      :arrowDirection="active ? 'up' : 'down'"
      @click="onClick"
    />
    <div
      class="ak-collapse-item_body ak-horizontal-hairline"
      style="display: none"
      ref="bodyEl"
    >
      <div class="ak-collapse-item_content">
        <slot></slot>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, ref, inject, computed } from 'vue'
import type { PropType } from 'vue'
import { Cell } from '../Cell'
import { iconValidator } from '../helpers/validator'
import { useGroupItem } from '../hooks/use-group'
import Exception from '../helpers/exception'
import type { CollapseItemEmits } from './types'
import type { PropsToEmits } from '../helpers/types'
import type { IconData } from '../Icon/types'
import { getItemClasses } from './util'
import { isBoolean } from '../helpers/util'

export default defineComponent({
  name: 'ak-collapse-item',
  components: { Cell },
  props: {
    icon: {
      type: [String, Object] as PropType<IconData>,
      validator: iconValidator
    },
    title: {
      type: String,
      default: ''
    },
    name: {
      type: [Number, String],
      default: '',
      required: true
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  emits: {
    toggle: payload => payload && isBoolean(payload.spread)
  } as PropsToEmits<CollapseItemEmits>,
  setup(props, { emit }) {
    const active = ref(false)
    const bodyEl = ref<HTMLElement>()
    const onChange = inject('akCollapseChange', collapseItemChange)
    const uid = Symbol()

    function collapseItemChange(uid: symbol) {
      new Exception(
        `CollapseItem is not in Collapse`,
        Exception.TYPE.DEFAULT,
        'CollapseItem'
      )
    }

    let visibleTimer: number

    function show(isClick = false) {
      if (active.value) {
        return
      }
      active.value = true

      clearTimeout(visibleTimer)

      const $body = bodyEl.value as HTMLElement

      $body.style.cssText = 'position: absolute; opacity: 0;'
      const contentHeight = $body.getBoundingClientRect().height
      $body.style.cssText = 'height: 0px;'

      visibleTimer = window.setTimeout(() => {
        $body.style.cssText = `height: ${contentHeight}px;`

        visibleTimer = window.setTimeout(() => {
          $body.style.cssText = ''
        }, 210)
      }, 17)

      emitToggle(true)

      isClick && onChange(uid)
    }

    function hide(isClick = false) {
      if (!active.value) {
        return
      }
      active.value = false

      clearTimeout(visibleTimer)

      const $body = bodyEl.value as HTMLElement
      $body.style.cssText = `height: ${$body.getBoundingClientRect().height}px;`

      visibleTimer = window.setTimeout(() => {
        $body.style.cssText = 'height: 0px;'

        visibleTimer = window.setTimeout(() => {
          $body.style.cssText = 'display: none;'
        }, 210)
      }, 17)

      emitToggle(false)

      isClick && onChange(uid)
    }

    function emitToggle(spread: boolean) {
      emit('toggle', {
        name: props.name,
        spread
      })
    }

    function onClick() {
      active.value ? hide(true) : show(true)
    }

    useGroupItem('collapse', {
      uid,
      getName: () => props.name,
      getActive: () => active.value,
      show,
      hide
    })

    const classes = computed(() => getItemClasses(active.value))

    return {
      active,
      bodyEl,
      onClick,
      classes
    }
  }
})
</script>
