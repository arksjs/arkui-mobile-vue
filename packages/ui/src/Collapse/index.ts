import { multiWithInstall } from '../helpers/with-install'
import Collapse from './Collapse.vue'
import CollapseItem from './CollapseItem.vue'

export { Collapse, CollapseItem }
export default Collapse
export const install = multiWithInstall(Collapse, [CollapseItem])
