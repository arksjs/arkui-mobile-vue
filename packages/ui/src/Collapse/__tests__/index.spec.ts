import { describe, test, expect } from 'vitest'
import { mount, shallowMount } from '@vue/test-utils'
import { Collapse, CollapseItem } from '@/Collapse'
import { baseList } from './data'
import CircleOutlined from '@/Icon/icons/CircleOutlined'
import { markRaw } from 'vue'

const CollapseTmpl = {
  template: `
  <ak-collapse v-model="modelValue" :accordion="accordion">
    <ak-collapse-item v-for="(item, index) in list" :key="item.title" :name="item.title" :title="item.title">
      {{ item.content }}
    </ak-collapse-item>
  </ak-collapse>
    `,
  components: { 'ak-collapse': Collapse, 'ak-collapse-item': CollapseItem },
  props: {
    accordion: Boolean
  },
  data() {
    return {
      modelValue: [],
      list: baseList
    }
  }
}

describe('Collapse', () => {
  test('snapshot', () => {
    const wrapper = shallowMount(Collapse, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('modelValue', async () => {
    const wrapper = mount(CollapseTmpl)

    expect(wrapper.vm.modelValue).toEqual([])

    // click item 1
    await wrapper.find('.ak-collapse-item_header').trigger('click')
    expect(wrapper.vm.modelValue).toEqual(['title 1'])

    // click item 2
    await wrapper.findAll('.ak-collapse-item_header')[1].trigger('click')
    expect(wrapper.vm.modelValue).toEqual(['title 1', 'title 2'])

    // click item 1
    await wrapper.find('.ak-collapse-item_header').trigger('click')
    expect(wrapper.vm.modelValue).toEqual(['title 2'])
  })

  test('accordion', async () => {
    const wrapper = mount(CollapseTmpl)
    await wrapper.setProps({
      accordion: true
    })

    expect(wrapper.vm.modelValue).toEqual([])

    // click item 1
    await wrapper.find('.ak-collapse-item_header').trigger('click')
    expect(wrapper.vm.modelValue).toEqual(['title 1'])

    // click item 2
    await wrapper.findAll('.ak-collapse-item_header')[1].trigger('click')
    expect(wrapper.vm.modelValue).toEqual(['title 2'])

    // click item 1
    await wrapper.find('.ak-collapse-item_header').trigger('click')
    expect(wrapper.vm.modelValue).toEqual(['title 1'])
  })
})

describe('CollapseItem', () => {
  const name = 'name'

  test('snapshot', () => {
    const wrapper = shallowMount(CollapseItem, {
      props: {
        name
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('title', () => {
    const title = 'title'
    const wrapper = mount(CollapseItem, {
      props: {
        name,
        title
      }
    })

    expect(wrapper.find('.ak-cell_label').text()).toBe(title)
  })

  test('icon', () => {
    const wrapper = mount(CollapseItem, {
      props: {
        name,
        icon: markRaw(CircleOutlined)
      }
    })

    expect(wrapper.findComponent(CircleOutlined).exists).toBeTruthy()
  })

  test('disabled', () => {
    const wrapper = mount(CollapseItem, {
      props: {
        name,
        disabled: true
      }
    })

    expect(wrapper.find('.ak-collapse-item_header').classes()).toContain(
      'disabled'
    )
  })
})
