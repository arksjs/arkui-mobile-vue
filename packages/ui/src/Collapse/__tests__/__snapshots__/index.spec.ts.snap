// Vitest Snapshot v1

exports[`Collapse > snapshot 1`] = `"<div class=\\"ak-collapse\\"></div>"`;

exports[`CollapseItem > snapshot 1`] = `
"<div class=\\"ak-collapse-item ak-horizontal-hairline\\">
  <cell-stub arrowdirection=\\"down\\" label=\\"\\" description=\\"\\" content=\\"\\" clickable=\\"false\\" required=\\"false\\" islink=\\"true\\" disabled=\\"false\\" class=\\"ak-collapse-item_header\\"></cell-stub>
  <div class=\\"ak-collapse-item_body ak-horizontal-hairline\\" style=\\"display: none;\\">
    <div class=\\"ak-collapse-item_content\\"></div>
  </div>
</div>"
`;
